### PR TITLE
WIP: CMake overhaul

### PR DESCRIPTION
**Pull request by friendlyanon**
- [x] I have read the contributing guidelines.

## Goal of this PR

CMake lists files are code and as such they should be treated with the same high expectations other code should be.  
As such, this PR aims to increase the maintainability and readability of the lists files by adapting modern CMake idioms.  
Additionally, it would be nice if consuming this library would be as simple in the common case as the following:
```cmake
include(FetchContent)
FetchContent_Declare(simdjson
        GIT_REPOSITORY ...
        GIT_TAG ...
        UPDATE_DISCONNECTED YES)
FetchContent_MakeAvailable(simdjson)

target_link_libraries(my_target PRIVATE simdjson::simdjson)
```

## How is the above achieved

The lists file in the root directory must not contain anything other than what is exclusivley a build or usage requirement. It should contain as little code as possible to make any changes or potential issues obvious, as this is the file that dependents will consume.  
This excludes from the top level lists file things like: tests, benchmarks, tools, examples, etc. These are not strictly part of the project, but they are tightly coupled and as such do not deserve their own repositories for logical reasons.  
Instead, separate directories must be created with their own "top level" lists files, that include the library's as a dependency (`FetchContent_Declare(simdjson SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/.."` is a nice and automatic way to do that, see the test lists file in the PR for example).

## Why is this PR marked `WIP:`?

I succeeded in cleaning out the top level lists file so far, but I'm not yet able to successfully build the library.  
As the quality of the lists file was previously of poor quality, I do not know what it is that I did wrong here, so I require input on what might be causing the issue.

## Questions

I tried to make my commits as atomic as possible, with extended reasoning in in commit messages where necessary, so I encourage anyone looking at this PR to first read thru the commit messages.  
If there are more questions of course do please ask them and I'll do my best to answer.

### CODE REVIEW

**Comment by pauldreik** [comment]
Seems like a good idea to improve the build config, but difficult to pull off.
How come you removed the submodules?

**Comment by friendlyanon** [comment]
> difficult to pull off

It's mostly just time consuming, not difficult.

> How come you removed the submodules?

Modern CMake provides utilities to handle 3rd party dependencies without having to rely on submodules tracked by git.  
Since CMake 3.14 one can very easily pull in 3rd parties:
```cmake
# Ideally the 3rd party named "foo" would have prefixed variables that we can
# set ahead of time to control some aspects of it
set(FOO_CHANGE_BEHAVIOUR YES CACHE INTERNAL "")

include(FetchContent)
FetchContent_Declare(foo
        GIT_REPOSITORY ...
        GIT_TAG ...
        UPDATE_DISCONNECTED YES)
FetchContent_MakeAvailable(foo)
```

**Comment by jkeiser** [comment]
I'd be happy to see more idiomatic cmake. I cobbled much of the current version together, but it was my first time using cmake :) It's been improved by others since then, but I expect most weirdness is mine.

I also find submodules mess with my repo (it keeps getting marked dirty), so I wouldn't mind getting to a place without them.

**Comment by friendlyanon** [comment]
I'm glad about the general positive reception for the PR, which in hindsight I should've first consulted with you via an issue. (Oops!)

However, as you can see the build is failing as of now, because I'm not completely familiar with the project and I don't know the complete build requirements and what targets this project should be exporting/making available for consumption.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,87 +1,89 @@
...
+        COMPONENT simdjson_Development
+        INCLUDES #
+        DESTINATION "${simdjson_include_directory}")
+
+set(simdjson_install_cmakedir
+        "${CMAKE_INSTALL_LIBDIR}/cmake/${simdjson_directory}")
+
+write_basic_package_version_file(
+        simdjsonConfigVersion.cmake
+        VERSION ${PROJECT_VERSION}
```
Comment (friendlyanon) [inline CMakeLists.txt:71]: Should `PROJECT_VERSION` be used here or the fake `SIMDJSON_LIB_VERSION`?

This argument here affects people importing the library with version requirements:
```cmake
find_package(simdjson 4) # will match 4.x.x.x if COMPATIBILITY is SameMajorVersion
```

Comment (lemire) [reply CMakeLists.txt:71]: Please do not change anything anything having to do with versions. We semantic versioning as well as library versioning, and a so version, and we have a scripted release process that updates automatically the files and CMake files, keeping everything in sync. All lines where versions are specify need to remain unchanged.

If you want to change how simdjson is versioned, please open a distinct issue, and let us discuss it. We won't change versioning without very deliberate and very thorough discussions.

We have a system that works currently, it meets the needs of our downstream users. We have a good release process. We don't want to break any of this.

How versioning works on the different systems we support is a complex set of conventions.

Please be mindful of our release process. Do not break it.

**Comment by lemire** [comment]
Please leave these lines unchanged...

```CMake
set(PROJECT_VERSION_MAJOR 0)
set(PROJECT_VERSION_MINOR 6)
set(PROJECT_VERSION_PATCH 0)
set(SIMDJSON_SEMANTIC_VERSION "0.6.0" CACHE STRING "simdjson semantic version")
set(SIMDJSON_LIB_VERSION "4.0.0" CACHE STRING "simdjson library version")
set(SIMDJSON_LIB_SOVERSION "4" CACHE STRING "simdjson library soversion")
```

**Comment by friendlyanon** [comment]
The variables are already declared by `project()`, see the link referenced in 45548161c89616f725f29fa677a19e46772a753b

**Comment by lemire** [comment]
The general rule that should guide us is: "if it ain't broke, don't fix it".

I agree that submodules are terrible. If there are better ways to do it, let us test it out and adopt them.

If you want to contribute a PR that removes submodules, and replace them with something better... You are totally correct that this will be well received.

If you can make some of our CMake code more idiomatic, that's great too... but...

A good strategy is to break your contributions into smaller PRs, each one with a clear focus. You do not get paid in terms of the number of lines that you change in one single PR. It is totally fine to issue 15 different PRs. We will be happy to receive them. There is no need to put everything into one giant PR that changes everything at once.

A PR that changes a single line is perfectly fine.

If you use small PRs, there is much greater chance that it will pass our tests, and that we will be able to think it through collectively.

When there are many, many things being changed, it is hard to be sure that nothing important is broken.

> The variables are already  (...)

Yes I know.

I recommend you break your PR into small ones. There are many things to improve... Pick one, one specific thing, and write a PR about this one particular issue. For example, if you want to remove the submodules, then do that, but only that.

**Comment by friendlyanon** [comment]
I see, very reasonable qualms regarding the reach of this PR.  
I admit I might be a little bit overzealous when it comes to CMake, which might not be the best attribute of mine when it comes to contributing to bigger projects, which I also lack experience with.

I'll begin breaking down what I have done so far, beginning with submodules as that seems to have gained everyone's attention especially.

> if it ain't broke, don't fix it

Yes, however it should also be pointed out that currently version numbers have to be updated in 7 places, which can be reduced to just 3, one of which can be used to automatically propagate the changes in the source files using CMake.

**Comment by lemire** [comment]
@friendlyanon +1

**Pull request closed without merging**